engine/language_client_go/main.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+/*
+#cgo LDFLAGS: ./lib/libhello.dylib -ldl
+#include <stdlib.h>
+#include <stdbool.h>
+#include "./lib/baml.h"
+
+extern void trigger_callback(uint32_t, bool, const char *);
+*/
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type StreamResult[Partial any, Final any] struct {
+	partial *Partial
+	final   *Final
+	error   error
+}
+
+func (result *StreamResult[Partial, Final]) Partial() Partial {
+	return *result.partial
+}
+
+func (result *StreamResult[Partial, Final]) Final() Final {
+	return *result.final
+}
+
+func (result *StreamResult[Partial, Final]) IsFinal() bool {
+	return result.final != nil
+}
+
+func (result *StreamResult[Partial, Final]) IsPartial() bool {
+	return result.partial != nil
+}
+
+func (result *StreamResult[Partial, Final]) Error() error {
+	return result.error
+}
+
+type DoSomethingStreamResult struct {
+	StreamResult[string, string]
+}
+
+// UpperCase in Go only!
+func DoSomethingStream(arg string) <-chan DoSomethingStreamResult {
+	// Do something with BAML runtime here!
+	// C.trigger_callback(C.uint(id), C.bool(false), C.CString(arg))
+
+	function_name := C.CString("TestOllama")
+
+	kwargs := C.CKwargs{
+		len:    0,
+		keys:   nil,
+		values: nil,
+	}
+
+	callback_id, callback := create_unique_id()
+
+	C.call_function_from_c(runtime, function_name, &kwargs, callback_id)
+
+	return_channel := make(chan DoSomethingStreamResult)
+	go func() {
+		for result := range callback {
+			// TODO: Handle the result
+			// error handling, type checking, etc.
+			return_channel <- DoSomethingStreamResult{StreamResult: StreamResult[string, string]{partial: &result.data}}
```

Comment:
This will block this goroutine and hence the trigger callback goroutine untill the user reads the message from this channel. This can cause the rust thread to block as well. I suggest you make the channel  on line 65 a buffered channel to avoid blocking on this line

Reply:
it doesn't look like its blocking! ran a test and it runs all 3 in parallel from what i see

Reply:
The user may not read the channel at the pace you are. What if the user decides to do something compute intensive in the range loop on the receiving end of this channel?

Reply:
Can confirm with an unbuffered channel this is a blocking send until the consumer reads.